awx/ui/client/src/instance-groups/instance-groups.service.js

```diff
@@ -9,33 +9,53 @@ export default
             },
             editInstanceGroups: function(url, instance_groups) {
                 Rest.setUrl(url);
-                let currentGroups = Rest.get()
-                    .then(({data}) => {
-                        return data.results.map((i) => i.id);
-                    });
+                return Rest.get()
+                    .then(res => {
+                        const { data: { results = [] } } = res;
+                        const updatedGroupIds = (instance_groups || []).map(({ id }) => id);
+                        const currentGroupIds = results.map(({ id }) => id);
 
-                return currentGroups.then(function(current) {
+                        const groupIdsToAssociate = [];
+                        const groupIdsToDisassociate = [];
 
-                    let groupsToAdd = (instance_groups || [])
-                        .map(val => val.id);
```

Comment:
compare the currently saved ids and the selected ids - as soon as we encounter a difference between the two arrays, we mark _all_ remaining currently saved ids for deletion.

Then, we build the association array in the correct order by adding any id that was marked for disassociation due to being out of order _and_ any ids that are new associations.

```
current:  [1,2,3,4,5,6]
selected: [1,2,4,5,6,3] 

-> idsToDisassociate = [3,4,5,6]
-> idsToAssociate = [2,3]
```